### PR TITLE
Ensure that register() is executed only once so that the resourcesQue…

### PR DESCRIPTION
**Pull request by verma7**
…ue does not become full

### CODE REVIEW

**Comment by verma7** [comment]
Fixes #372.

**Comment by verma7** [comment]
retest this please

**Comment by cooldoger** [comment]
Looks good to me.

**Comment by vargup** [comment]
Code looks good to me, but do we know the reason, why register method is called twice?

**Comment by verma7** [comment]
@vargup Yeah, that's the contract with Mesos Scheduler API: https://issues.apache.org/jira/browse/MESOS-786

```
1) Framework connects w/ master very first time --> registered()
2) Framework reconnects w/ same master after a zk blip --> reregistered()
3) Framework reconnects w/ failed over master --> registered()
4) Failed over framework connects w/ same master --> registered()
5) Failed over framework connects w/ failed over master --> registered()
```

File: cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/CassandraScheduler.java [hunk 1]
```diff
@@ -141,6 +142,12 @@ public CassandraScheduler(
     public void registered(SchedulerDriver driver,
                            Protos.FrameworkID frameworkId,
                            Protos.MasterInfo masterInfo) {
+        // Return early if registered() has already been called once in the lifetime of this object.
+        if (!isSchedulerRegistered.compareAndSet(false, true)) {
```
Comment (mohitsoni) [inline cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/CassandraScheduler.java]: @verma7 Just curious here. As per point 3 in this:
```
1) Framework connects w/ master very first time --> registered()
2) Framework reconnects w/ same master after a zk blip --> reregistered()
3) Framework reconnects w/ failed over master --> registered()
4) Failed over framework connects w/ same master --> registered()
5) Failed over framework connects w/ failed over master --> registered()
```

Shouldn't the reconciler triggered again ? If yes, we can prevent re-init of objects but can still perform reconciliation.

Comment (verma7) [inline cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/CassandraScheduler.java]: The reconciler should already be started on L173 and reconcile() is called every time we receive an offer. Do you want me to call reconciler.start() again? Or reconcile() ?

Comment (verma7) [inline cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/CassandraScheduler.java]: I am inclined to just call reregistered() in this case. Does that sound ok @mohitsoni ?

Comment (mohitsoni) [inline cassandra-scheduler/src/main/java/com/mesosphere/dcos/cassandra/scheduler/CassandraScheduler.java]: I also think that calling `reregistered()` should be sufficient in this case.

Also, you are right that reconcile() is called every time we receive an offer. But in a busy cluster, offer cycle can either be delayed OR in worst case the framework might not receive offer for a long time. And, for that duration the framework will live the lie about it's view of world for a longer time v/s the pro-active reconcile() triggered as part of re-registered call that you proposed.

What do you think ?

**Comment by verma7** [comment]
retest this please

**Comment by verma7** [comment]
retest this please

**Comment by verma7** [comment]
@gabrielhartmann @mohitsoni Can you please tell me why the tests are failing?

**Comment by verma7** [comment]
@gabrielhartmann @mohitsoni Can I merge this PR?

**Comment by verma7** [comment]
@nickbp Can you please merge this?

**Pull request merged**